rust/connlib/tunnel/src/sockets.rs

```diff
@@ -191,20 +192,49 @@ impl<const N: usize> Socket<N> {
     }
 }
 
-fn make_socket(addr: impl Into<SocketAddr>) -> io::Result<std::net::UdpSocket> {
+fn make_socket(
+    addr: impl Into<SocketAddr>,
+    callbacks: &CallbackErrorFacade<impl Callbacks>,
+) -> Result<std::net::UdpSocket> {
     let addr: SockAddr = addr.into().into();
     let socket = socket2::Socket::new(addr.domain(), Type::DGRAM, None)?;
+    protect(&socket, callbacks)?;
     socket.set_nonblocking(true)?;
     socket.bind(&addr)?;
 
-    // TODO: for android protect file descriptor
-    #[cfg(target_os = "linux")]
-    socket.set_mark(crate::FIREZONE_MARK)?;
-
     // Note: for AF_INET sockets IPV6_V6ONLY is not a valid flag
     if addr.is_ipv6() {
         socket.set_only_v6(true)?;
     }
 
     Ok(socket.into())
 }
+
+#[cfg(target_os = "android")]
+fn protect(
+    socket: &socket2::Socket,
+    callbacks: &CallbackErrorFacade<impl Callbacks>,
+) -> Result<()> {
+    use connlib_shared::Callbacks;
+    use std::os::fd::AsRawFd;
+
+    callbacks.protect_file_descriptor(socket.as_raw_fd())?;
```

Comment:
Is this strictly necessary? We didn't do that before and I have some reservations about introducing more callbacks :)

See https://github.com/firezone/firezone/issues/3458.

Reply:
If it is necessary, I think it would be cleaner to (for now) instead have getters on `Tunnel` to expose the file descriptors in order for the android client code to call `protect` on it.

At a later point, we can then explore doing something like what I outlined in #3458. At least, that would avoid passing the callbacks down to low-level code like this.

Another alternative would be to make the `Sockets` an argument of the tunnel. That way, the android glue code could configure those before constructing the tunnel.

Reply:
Let me know what you think of [5973943](https://github.com/firezone/firezone/pull/3391/commits/5973943c3d1ddb62a1b0b2a8f478d304cfc917fe).